File: Soluciones_Ejercicios/Java/Ejercicio2.java
import java.util.LinkedList;
import java.util.Queue;

public class Ejercicio2 {

    static class Nodo {
        int dato;
        Nodo izq, der;
        Nodo(int v) { dato = v; }
    }

    static Nodo insertar(Nodo raiz, int val) {
        if (raiz == null) return new Nodo(val);
        if (val < raiz.dato) raiz.izq = insertar(raiz.izq, val);
        else                 raiz.der = insertar(raiz.der, val);
        return raiz;
    }

    static void preorden(Nodo r)  { if (r==null) return; System.out.print(r.dato+" "); preorden(r.izq); preorden(r.der); }
    static void inorden(Nodo r)   { if (r==null) return; inorden(r.izq); System.out.print(r.dato+" "); inorden(r.der); }
    static void postorden(Nodo r) { if (r==null) return; postorden(r.izq); postorden(r.der); System.out.print(r.dato+" "); }

    static void bfs(Nodo raiz) {
        if (raiz == null) return;
        Queue<Nodo> q = new LinkedList<>();
        q.add(raiz);
        while (!q.isEmpty()) {
            Nodo a = q.poll();
            System.out.print(a.dato + " ");
            if (a.izq != null) q.add(a.izq);
            if (a.der != null) q.add(a.der);
        }
    }

    public static void main(String[] args) {
        Nodo raiz = null;
        // Arbol original
        for (int v : new int[]{10, 5, 15, 2, 7, 12, 20})
            raiz = insertar(raiz, v);
        // Nuevos nodos: 1, 3, 18, 25
        for (int v : new int[]{1, 3, 18, 25})
            raiz = insertar(raiz, v);

        System.out.print("Preorden:  "); preorden(raiz);  System.out.println();
        System.out.print("Inorden:   "); inorden(raiz);   System.out.println();
        System.out.print("Postorden: "); postorden(raiz); System.out.println();
        System.out.print("BFS:       "); bfs(raiz);       System.out.println();
    }
}

File: Soluciones_Ejercicios/C++/Ejercicio2.cpp
#include <iostream>
#include <queue>
using namespace std;

struct Nodo {
    int dato;
    Nodo* izq;
    Nodo* der;
    Nodo(int v) : dato(v), izq(nullptr), der(nullptr) {}
};

Nodo* insertar(Nodo* raiz, int val) {
    if (!raiz) return new Nodo(val);
    if (val < raiz->dato) raiz->izq = insertar(raiz->izq, val);
    else                  raiz->der = insertar(raiz->der, val);
    return raiz;
}

void preorden(Nodo* r)  { if (!r) return; cout << r->dato << " "; preorden(r->izq); preorden(r->der); }
void inorden(Nodo* r)   { if (!r) return; inorden(r->izq); cout << r->dato << " "; inorden(r->der); }
void postorden(Nodo* r) { if (!r) return; postorden(r->izq); postorden(r->der); cout << r->dato << " "; }

void bfs(Nodo* raiz) {
    if (!raiz) return;
    queue<Nodo*> q;
    q.push(raiz);
    while (!q.empty()) {
        Nodo* a = q.front(); q.pop();
        cout << a->dato << " ";
        if (a->izq) q.push(a->izq);
        if (a->der) q.push(a->der);
    }
}

int main() {
    Nodo* raiz = nullptr;
    // Arbol original
    for (int v : {10, 5, 15, 2, 7, 12, 20})
        raiz = insertar(raiz, v);
    // Nuevos nodos: 1, 3, 18, 25
    for (int v : {1, 3, 18, 25})
        raiz = insertar(raiz, v);

    cout << "Preorden:  "; preorden(raiz);  cout << endl;
    cout << "Inorden:   "; inorden(raiz);   cout << endl;
    cout << "Postorden: "; postorden(raiz); cout << endl;
    cout << "BFS:       "; bfs(raiz);       cout << endl;
    return 0;
}
